google/cloud/internal/future_impl.h

```diff
@@ -0,0 +1,198 @@
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_FUTURE_IMPL_H_
+/**
+ * @file
+ *
+ * Define the implementation details for `google::cloud::future<T>`.
+ */
+
+#include "google/cloud/internal/port_platform.h"
+
+// C++ futures only make sense when exceptions are enabled.
+#if GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
+#include "google/cloud/version.h"
+#include <condition_variable>
+#include <exception>
+#include <future>
+#include <mutex>
+
+namespace google {
+namespace cloud {
+inline namespace GOOGLE_CLOUD_CPP_NS {
+namespace internal {
+/**
+ * Common base class for all shared state classes.
+ *
+ * The implementation of the shared state for future<T>, future<R&> and
+ * future<void> share a lot of code. This class refactors that code, it
+ * represents a shared state of unknown type.
+ *
+ * @note While most of the invariants for promises and futures are implemented
+ *   by this class, not all of them are. Notably, future values can only be
+ *   retrieved once, but this is enforced because calling .get() or .then() on a
+ *   future invalidates the future for further use. The shared state does not
+ *   record that state change.
+ */
+class future_shared_state_base {
+ public:
+  future_shared_state_base() : mu_(), cv_(), current_state_(state::not_ready) {}
+
+  /// Return true if the shared state has a value or an exception.
+  bool is_ready() const {
+    std::unique_lock<std::mutex> lk(mu_);
+    return is_ready_unlocked();
+  }
+
+  /// Block until is_ready() returns true ...
+  void wait() {
+    std::unique_lock<std::mutex> lk(mu_);
+    cv_.wait(lk, [this] { return is_ready_unlocked(); });
+  }
+
+  /**
+   * Block until `is_ready()` returns true or until @p duration time has
+   * elapsed.
+   *
+   * @param duration the maximum time to wait for `is_ready()`.
+   *
+   * @tparam Rep a placeholder to match the Rep tparam for @p duration's
+   *     type, the semantics of this template parameter are documented in
+   *     `std::chrono::duration<>` (in brief, the underlying arithmetic type
+   *     used to store the number of ticks), for our purposes it is simply a
+   *     formal parameter.
+   * @tparam Period a placeholder to match the Period tparam for @p duration's
+   *     type, the semantics of this template parameter are documented in
+   *     `std::chrono::duration<>` (in brief, the length of the tick in seconds,
+   *     expressed as a `std::ratio<>`), for our purposes it is simply a formal
+   *     parameter.
+   *
+   * @return `std::future_status::ready` if the shared state is satisfied.
+   *     `std::future_status::deferred` if the shared state is not satisfied and
+   *     there is a continuation ready to execute when it is satisfied.
+   *     `std::future_status::timeout` otherwise.
+   */
+  template <typename Rep, typename Period>
+  std::future_status wait_for(std::chrono::duration<Rep, Period> duration) {
+    std::unique_lock<std::mutex> lk(mu_);
+    if (not lk.owns_lock()) {
```

Comment:
Why this `if` necessary? If I'm reading the docs and code correctly, `lk` will acquire the mutex. Same question for `wait_until`.